File: assignments/rust/word-count/word-count_test.rs
#[crate_id = "word-count_test#1.0"];
#[crate_type = "lib"];

use std::hashmap::HashMap;

mod word_count;

fn check_word_count(s: &str, pairs: &[(&str, uint)]) {
    // The reason for the awkward code in here is to ensure that the failure
    // message for assert_eq! is as informative as possible. A simpler
    // solution would simply check the length of the map, and then
    // check for the presence and value of each key in the given pairs vector.
    let mut m: HashMap<~str, uint> = word_count::word_count(s);
    for &(k, v) in pairs.iter() {
        assert_eq!((k, m.pop(&k.to_owned()).unwrap()), (k, v));
    }
    // may fail with a message that clearly shows all extra pairs in the map
    assert_eq!(m.iter().to_owned_vec(), ~[]);
}

#[test]
#[ignore]
fn test_no_words() {
    check_word_count("", []);
    check_word_count("-!@#$ $#() @@@", []);
}

#[test]
#[ignore]
fn test_count_one_word() {
    check_word_count("word", [("word", 1)]);
}

#[test]
#[ignore]
fn test_count_one_of_each() {
    check_word_count(
        "one of each",
        [("one", 1),
         ("of", 1),
         ("each", 1)]);
}

#[test]
#[ignore]
fn test_count_multiple_occurrences() {
    check_word_count(
        "one fish two fish red fish blue fish",
        [("one", 1),
         ("fish", 4),
         ("two", 1),
         ("red", 1),
         ("blue", 1)]);
}

#[test]
#[ignore]
fn test_ignore_punctuation() {
    check_word_count(
        "car : carpet as java : javascript!!&@$%^&",
        [("car", 1),
         ("carpet", 1),
         ("as", 1),
         ("java", 1),
         ("javascript", 1)]);
}

#[test]
#[ignore]
fn test_include_numbers() {
    check_word_count(
        "testing, 1, 2 testing",
        [("testing", 2),
         ("1", 1),
         ("2", 1)]);
}

#[test]
#[ignore]
fn test_normalize_case() {
    check_word_count(
        "go Go GO",
        [("go", 3)]);
}

#[test]
#[ignore]
fn test_prefix_punctuation() {
    check_word_count(
        "!%%#testing, 1, 2 testing",
        [("testing", 2),
         ("1", 1),
         ("2", 1)]);
}

#[test]
#[ignore]
fn test_symbols_are_separators() {
    check_word_count(
        "hey,my_spacebar_is_broken.",
        [("hey", 1),
         ("my", 1),
         ("spacebar", 1),
         ("is", 1),
         ("broken", 1)]);
}

File: assignments/go/word-count/word-count_test.go
package wc

import "testing"

func assertEqual(t *testing.T, actual Histogram, expected Histogram) {
	if !actual.Equal(expected) {
		t.Errorf("Expected %v to equal %v", actual, expected)
	}
}

func assertNotEqual(t *testing.T, h Histogram, other Histogram) {
	if h.Equal(other) || other.Equal(h) {
		t.Errorf("Expected %v to not equal %v", h, other)
	}
}

func TestEqual(t *testing.T) {
	h1 := Histogram{"hello": 1, "world": 2}
	h2 := Histogram{"hello": 1, "world": 2}
	assertEqual(t, h1, h2)
}

func TestNotEqual(t *testing.T) {
	t.SkipNow()
	h1 := Histogram{"word": 1}
	h2 := Histogram{"word": 1, "games": 2}
	assertNotEqual(t, h1, h2)
}

func TestCountOneWord(t *testing.T) {
	t.SkipNow()
	h := Histogram{"word": 1}
	assertEqual(t, WordCount("word"), h)
}

func TestCountOneOfEach(t *testing.T) {
	t.SkipNow()
	h := Histogram{"one": 1, "of": 1, "each": 1}
	assertEqual(t, WordCount("one of each"), h)
}

func TestCountMultipleOccurrences(t *testing.T) {
	t.SkipNow()
	actual := WordCount("one fish two fish red fish blue fish")
	expected := Histogram{"one": 1, "fish": 4, "two": 1, "red": 1, "blue": 1}
	assertEqual(t, actual, expected)
}

func TestIgnorePunctuation(t *testing.T) {
	t.SkipNow()
	actual := WordCount("car : carpet as java : javascript!!&@$%^&")
	expected := Histogram{"car": 1, "carpet": 1, "as": 1, "java": 1, "javascript": 1}
	assertEqual(t, actual, expected)
}

func TestIncludeNumbers(t *testing.T) {
	t.SkipNow()
	actual := WordCount("testing, 1, 2 testing")
	expected := Histogram{"testing": 2, "1": 1, "2": 1}
	assertEqual(t, actual, expected)
}

func TestNormalizeCase(t *testing.T) {
	t.SkipNow()
	assertEqual(t, WordCount("go Go GO"), Histogram{"go": 3})
}
